src/main/java/uk/ac/cam/cl/dtg/segue/database/GitDb.java

```diff
@@ -141,19 +141,55 @@ public ByteArrayOutputStream getFileByCommitSHA(final String sha, final String f
             return null;
         }
 
-        ObjectId objectId = this.findGitObject(sha, fullFilePath);
+        Repository repository = gitHandle.getRepository();
+        // This may or may not help with concurrent repo update issues:
+        repository.scanForRepoChanges();
+
+        ObjectId commitId = repository.resolve(sha);
+
+        RevWalk revWalk = new RevWalk(repository);
+        RevCommit commit = revWalk.parseCommit(commitId);
+
+        RevTree tree = commit.getTree();
+
+        TreeWalk treeWalk = new TreeWalk(repository);
+        treeWalk.addTree(tree);
+        treeWalk.setRecursive(true);
+        treeWalk.setFilter(PathFilter.create(fullFilePath));
+
+        int count = 0;
+        ObjectId objectId = null;
+        String path = null;
+        while (treeWalk.next()) {
+            count++;
+            if (null == objectId) {
+                objectId = treeWalk.getObjectId(0);
+                path = treeWalk.getPathString();
+            } else if (count > 1) {
+                // throw exception if we find that there is more than one that matches the search.
+                StringBuilder sb = new StringBuilder();
+                sb.append("Multiple results have been found in the git repository for the following search: ");
+                sb.append(fullFilePath).append(".");
+                sb.append(" in ");
+                sb.append(sha);
+                sb.append(" Unable to decide which one to return.");
+                throw new UnsupportedOperationException(sb.toString());
+            }
+        }
 
         if (null == objectId) {
             return null;
         }
 
-        Repository repository = gitHandle.getRepository();
+        revWalk.dispose();
+        log.debug("Retrieved Commit Id: " + commitId.getName() + " Searching for: " + fullFilePath + " found: " + path);
         ObjectLoader loader = repository.open(objectId);
 
         ByteArrayOutputStream out = new ByteArrayOutputStream();
         loader.copyTo(out);
 
-        repository.close();
+        // TODO: Calling close seems to be unnecessary when not writing to the repo, and prints an error frequently.
+        //repository.close();
```

Comment:
As you noted in the tech meeting, we reuse the repo and so this is not something we need to do.
Close seems to just clear caches which is actually not something that would help so let's go with it.

Reply:
I am still a little worried about it, but the documentation has no examples like our use-case (a long-running multi-threaded environment) beyond noting that JGit is not generally thread safe except for read operations (which is all we do).
Let's try it and see if this introduces more errors. It might be that closing it prematurely was the reason the patch files are out-of-date, even 🤷